lomond/parser.py

```diff
@@ -121,76 +128,86 @@ def feed(self, data):
                 self._awaiting = self._gen.throw(error)
 
+        if self._exhausted:
+            raise ParseOverflow("extra bytes in feed(); {!r}".format(data[:100]))
         if self._eof:
-            raise ParseError('end of file reached')
+            raise ParseEOF(
+                "end of file reached; feed() has previously been called with empty bytes"
+            )
         if not data:
             self._eof = True
-            self._gen.throw(
-                ParseError('unexpected eof of file')
-            )
-
-        _buffer = self._buffer
-        pos = 0
-        while pos < len(data):
-            # Awaiting a read of a fixed number of bytes
-            if isinstance(self._awaiting, _ReadBytes):
-                # This many bytes left to read
-                remaining = self._awaiting.remaining
-                # Bite off remaining bytes
-                chunk = data[pos:pos + remaining]
-                chunk_size = len(chunk)
-                pos += chunk_size
-                try:
-                    # Validate new data
-                    self._awaiting.validate(chunk)
-                except ParseError as error:
-                    # Raises an exception in parse()
-                    self._awaiting = self._gen.throw(error)
-                # Add to buffer
-                _buffer.extend(chunk)
-                remaining -= chunk_size
-                if remaining:
-                    # Await more bytes
-                    self._awaiting.remaining = remaining
-                else:
-                    # Send to coroutine, get new 'awaitable'
-                    self._awaiting = self._gen.send(_buffer[:])
-                    del _buffer[:]
-
-            # Awaiting a read until a terminator
-            elif isinstance(self._awaiting, _ReadUntil):
-                # Reading to separator
-                chunk = data[pos:]
-                _buffer.extend(chunk)
-                sep = self._awaiting.sep
-                sep_index = _buffer.find(sep)
-
-                if sep_index == -1:
-                    # Separator not found, advance position
-                    pos += len(chunk)
-                    _check_length(len(_buffer))
-                else:
-                    # Found separator
-                    # Get data prior to and including separator
-                    sep_index += len(sep)
-                    _check_length(sep_index)
-                    # Reset data, to continue parsing
-                    data = _buffer[sep_index:]
-                    pos = 0
-                    # Send bytes to coroutine, get new 'awaitable'
-                    self._awaiting = self._gen.send(_buffer[:sep_index])
-                    del _buffer[:]
-
-            # Yield any non-awaitables...
-            while not isinstance(self._awaiting, _Awaitable):
-                yield self._awaiting
-                self._awaiting = next(self._gen)
+            self._gen.throw(ParseEOF("unexpected eof of file"))
```

Comment:
This function is too weighty. As some point I will break it up a little. It's on my list.